### ISSUE TITLE
Panic when checking pragma BUILTIN SHARP

### ISSUE THREAD

**Issue by berndlosert** [agda/agda#5610]
Agda panics when type-checking this code:

```agda
{-# OPTIONS --type-in-type #-}

infix 1000 ♯_

postulate
  ∞  : ∀ {a} (A : Set a) → Set a
  ♯_ : ∀ {a} {A : Set a} → A → ∞ A
  ♭  : ∀ {a} {A : Set a} → ∞ A → A

{-# BUILTIN INFINITY ∞  #-}
{-# BUILTIN SHARP    ♯_ #-}
{-# BUILTIN FLAT     ♭  #-}
```
The error is:
```
Panic: Pattern match failure in do expression at
src/full/Agda/TypeChecking/Rules/Builtin/Coinduction.hs:68:5-13
when checking the pragma BUILTIN SHARP ♯_
```

If I remove `{-# OPTIONS --type-in-type #-}`, it works again. I'm not sure if this is a bug or intended behavior.

**Comment by berndlosert** [comment]
I'm using 2.6.2 by the way.

**Comment by andreasabel** [comment]
Thanks for the report @berndlosert!  
Still works with 2.6.0, first breaks with 2.6.1.

I think the error location is still accurate:
https://github.com/agda/agda/blob/d8145b921a95c659d0313c1b84500d73081572cc/src/full/Agda/TypeChecking/Rules/Builtin/Coinduction.hs#L68

**Comment by andreasabel** [comment]
Bisection says: 7807e9a3920b3af24c5d576c9f7b887f64c9c539 is the first bad commit
Date:   Tue Oct 15 14:16:10 2019 +0200
[ fix #3073 ] Don't eta-contract `\x -> f l` to `f` when --type-in-type
Ping @jespercockx 

This looks plausible, because eta-contraction could have produced the `Def` form that is matched against at the problematic position.

**Comment by andreasabel** [comment]
The proposed fix 74f92b54a328a76221f72e3fb5b925bf0846c677 can be directly exploited, see https://github.com/agda/agda/pull/5614#issuecomment-950552130.
